src/Interpreters/OptimizeShardingKeyRewriteInVisitor.cpp

```diff
@@ -46,6 +46,9 @@ bool shardContains(
         return false;
 
     Field sharding_value = executeFunctionOnField(sharding_column_value, sharding_column_name, expr, sharding_key_column_name);
+    if (!isInt64FieldType(sharding_value.getType()))
```

Comment:
Maybe add a comment?

If sharding expression returned something different than UInt64 or Int64 (but why does it happen?) we leave the shard as possible containing this value (but why?).

Reply:
Is it possible because filtering happens on AST level before types are checked?
Then maybe we need to remove this shard rather than keeping it?

Reply:
>If sharding expression returned something different than UInt64 or Int64 (but why does it happen?)

The reason is that IN may contain other types, but they should be convertible to numeric types, I've just added type conversion